### PR TITLE
Updated examples and README.md

### PR DESCRIPTION
**Pull request by i5hi**
- Added examples/psbt.rs

- Updated error message for examples/bip32.rs.

The previous error would print the entire path as follows:
```
not enough arguments. usage: /home/ishi/rust-bitcoin/target/debug/examples/bip32 <WIF>
```

- Updated README.md with instructions to build and run examples.

### CODE REVIEW

File: examples/psbt.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+use std::{env, process};
+
+use bitcoin::util::psbt::PartiallySignedTransaction;
+use bitcoin::consensus::encode::deserialize;
+
+fn main() {
+
+  let args: Vec<String> = env::args().collect();
+  if args.len() < 2 {
+      eprintln!("not enough arguments. usage: cargo run --example psbt <PSBT>");
```
Comment (tcharding) [inline examples/psbt.rs:12]:
```suggestion
      eprintln!("Missing arguments. Usage: cargo run --example psbt <PSBT>");
```

File: examples/psbt.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      process::exit(1);
+  }
+  let decoded_psbt = match base64::decode(&args[1]){
```
Comment (tcharding) [inline examples/psbt.rs:15]:
```suggestion
  let decoded_psbt = match base64::decode(&args[1]) {
```
This is why we need a formatter :)

File: examples/psbt.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  let psbt_struct: PartiallySignedTransaction = match deserialize(&decoded_psbt){
+    Ok(psbt)=>psbt,
+    Err(_)=>{
+      eprintln!("Error deserialising psbt");
+      process::exit(1);
+    }
+  };
+
+  println!("{:#?}", &psbt_struct);
+
```
Comment (tcharding) [inline examples/psbt.rs:32]: I don't rekon we need this whitespace, or that at the start of `main()`. Damn this feels like reviewing C code ...

File: Cargo.toml [hunk 4]
```diff
@@ -50,10 +50,14 @@ secp256k1 = { version = "0.20.0", features = [ "recovery", "rand-std" ] }
 bincode = "1.3.1"
 # We need to pin ryu (transitive dep from serde_json) to stay compatible with Rust 1.22.0
 ryu = "<1.0.5"
+base64-compat = "1.0.0"
```
Comment (tcharding) [inline Cargo.toml:53]: Why is this here? I don't see it used anywhere in the PR?

Comment (i5hi) [reply Cargo.toml:53]: The examples/psbt.rs file needed it to use base64::decode.

File: examples/psbt.rs (hunk 2 above)
Comment (i5hi) [reply examples/psbt.rs:15]: damn, clippy results:
```
error: aborting due to previous error; 113 warnings emitted
```
I'll attempt addressing this in another PR.

File: examples/psbt.rs (hunk 3 above)
Comment (i5hi) [reply examples/psbt.rs]: :p sorted.

File: Cargo.toml (hunk 4 above)
Comment (tcharding) [reply Cargo.toml:53]: Oh nice, I didn't click that `base64-compat` had the lib name set to `base64`. Thanks

**Review by tcharding — APPROVED** [review]

**Comment by mcroad** [comment]
Having the same issue with rust-bitcoin/rust-miniscript#277.

To downgrade to 1.29 run `rustup default 1.29.0` and `rustc --version` to confirm.

**Comment by i5hi** [comment]
@mcroad

**Comment by i5hi** [comment]
@mcroad I did that. But that leads to a bunch of other errors.
Does just that command alone work for you to be able to run contrib/test.sh?

**Comment by i5hi** [comment]
@mcroad Just saw your issue on rust-miniscript. I get the same error here but with the cc crate. Will try out a few things tomorrow and get back.

**Comment by i5hi** [comment]
Aww man :/ Thats awkward :P Last time I attemtped this, I did run those pin commands and still got errors.

Noticed that if I ran `rustup default 1.29.0` then the pin commands and then `rustup default 1.29.0` again, it would error. I suppose `rustup default` must ONLY be run before pinning, running it both before and after (or just after) seems to reset the pin commands.

**Comment by i5hi** [comment]
All good now. Cheers @mcroad!

**Comment by i5hi** [comment]
Final changes to this PR are complete.

- psbt example works and is compatible with MSRV
- updated README.md with a few more pointers on testing changings with MSRV
- added psbt example to contrib/test.sh

**Comment by i5hi** [comment]
Added --features base64 to the new psbt example in contrib/test.sh which was failing the last CI run.

File: README.md [hunk 5]
```diff
@@ -61,6 +61,10 @@ please join us in
 
 This library should always compile with any combination of features on **Rust 1.29**.
 
+```
+rustup default 1.29.0
```
Comment (dr-orlovsky) [inline README.md:65]: Not sure whether its good to ask user to change the default rust version - especially with the fact that it will compile with any version above that

File: README.md [hunk 6]
```diff
@@ -79,6 +83,8 @@ For the feature `base64` to work with 1.29.0 we also need to pin `byteorder`:
 cargo update -p byteorder --precise "1.3.4"
 ```
 
+*If you face a `Cargo.lock` parsing error, remove it before running the above commands.*
```
Comment (dr-orlovsky) [inline README.md:86]: Or do `cargo generate-lockfile`

File: examples/psbt.rs (hunk 2 above)
Comment (dr-orlovsky) [reply examples/psbt.rs:18]: Now we have clippy mfixups erged so it should work for now. But also pls format this specific file with `rustfmt`

File: examples/psbt.rs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  let decoded_psbt = match base64::decode(&args[1]){
+    Ok(psbt)=> psbt,
+    Err(_)=>{
+      eprintln!("Error decoding psbt from base64");
+      process::exit(1);
+    }
+  };
+
+  let psbt_struct: PartiallySignedTransaction = match deserialize(&decoded_psbt){
```
Comment (dr-orlovsky) [inline examples/psbt.rs:26]: Nit:
```suggestion
  let psbt_struct: PartiallySignedTransaction = match deserialize(&decoded_psbt) {
```

**Review by dr-orlovsky — CHANGES_REQUESTED** [review]
Thank you and congrats with a first-time contribution!

I left few code nits and changes to README. Unfortunately you need to rebase on top of master and remove merge commit - we do not accept from-master merges in the history. Also, it is required that each of the commits compile successfully. Sorry for inconvenience, but this is a maintainer requirement applied to all PRs in this repo.

**Comment by i5hi** [comment]
Thanks for the support @dr-orlovsky 

Will fix this up to meet all the mentioned requirements and also address the nits you've mentioned.

File: README.md (hunk 5 above)
Comment (i5hi) [reply README.md:65]: I added this to help a contributor test with MSRV. Maybe add a note saying "To test with MSRV..." ?

Comment (dr-orlovsky) [reply README.md:65]: For testing I would prefer something like `rustup install 1.29.0 && cargo +1.29.0 build`, which will not switch the global rust environment

Comment (i5hi) [reply README.md:65]: Right, just `default` subcommand also  assumes 1.29.0 is already installed.

**Comment by tcharding** [comment]
Hi @i5hi, are you working on this PR still mate?

**Comment by i5hi** [comment]
Hey. Yes.

I will re-review, update and try to close it by the end of November.

**Comment by tcharding** [comment]
Awesome, thanks man!

**Comment by i5hi** [comment]
oh snap, i broke it while trying to sync up.

Will make a new PR.
